### PR TITLE
IGNITE-5580

### PR DESCRIPTION
**Pull request by SomeFire**

### CODE REVIEW

File: modules/core/src/main/java/org/apache/ignite/spi/discovery/tcp/TcpDiscoverySpi.java [hunk 1]
```diff
@@ -942,6 +978,35 @@ public TcpDiscoverySpi setTopHistorySize(int topHistSize) {
         return this;
     }
 
+    /**
+     * @return Size of topology snapshots history.
+     */
+    public long getEvtHistorySize() {
```
Comment (NSAmelchev) [inline modules/core/src/main/java/org/apache/ignite/spi/discovery/tcp/TcpDiscoverySpi.java]: int

File: modules/core/src/main/java/org/apache/ignite/spi/discovery/tcp/TcpDiscoverySpi.java [hunk 2]
```diff
@@ -268,6 +272,9 @@
     /** Maximum ack timeout value for receiving message acknowledgement in milliseconds (value is <tt>600,000ms</tt>). */
     public static final long DFLT_MAX_ACK_TIMEOUT = 10 * 60 * 1000;
 
+    /** Default size of discovery events history. */
+    public static final int DFLT_EVT_HISTORY_SIZE = 10;
```
Comment (NSAmelchev) [inline modules/core/src/main/java/org/apache/ignite/spi/discovery/tcp/TcpDiscoverySpi.java]: may be 100

File: modules/core/src/main/java/org/apache/ignite/spi/discovery/tcp/TcpDiscoverySpi.java [hunk 3]
```diff
@@ -402,6 +409,12 @@
     /** */
     protected IgniteSpiContext spiCtx;
 
+    /** Size of events history. */
+    private int evtHistSize = DFLT_EVT_HISTORY_SIZE;
+
+    /** Events history.*/
```
Comment (NSAmelchev) [inline modules/core/src/main/java/org/apache/ignite/spi/discovery/tcp/TcpDiscoverySpi.java]: whitespace

File: modules/core/src/main/java/org/apache/ignite/spi/discovery/tcp/TcpDiscoverySpi.java (hunk 1 above)
Comment (SomeFire) [reply modules/core/src/main/java/org/apache/ignite/spi/discovery/tcp/TcpDiscoverySpi.java]: Done.

File: modules/core/src/main/java/org/apache/ignite/spi/discovery/tcp/TcpDiscoverySpi.java (hunk 3 above)
Comment (SomeFire) [reply modules/core/src/main/java/org/apache/ignite/spi/discovery/tcp/TcpDiscoverySpi.java]: Done.

File: modules/core/src/main/java/org/apache/ignite/spi/discovery/tcp/TcpDiscoverySpi.java (hunk 2 above)
Comment (SomeFire) [reply modules/core/src/main/java/org/apache/ignite/spi/discovery/tcp/TcpDiscoverySpi.java]: Done.

File: modules/core/src/main/java/org/apache/ignite/spi/discovery/tcp/ServerImpl.java [hunk 4]
```diff
@@ -2070,6 +2073,27 @@ private void processMessageFailedNodes(TcpDiscoveryAbstractMessage msg) {
...
     }
 
+    /**
+     * Creates {@link TcpDiscoveryNodeFailedMessage} with specified message. This message will be showed on other nodes.
+     *
+     * @param failedNode Failed node.
+     * @param warningMsg Message to show on other nodes.
+     * @return Message about failed node.
+     */
+    private TcpDiscoveryNodeFailedMessage newTcpDiscoveryNodeFailedMessage(TcpDiscoveryNode failedNode,
```
Comment (NSAmelchev) [inline modules/core/src/main/java/org/apache/ignite/spi/discovery/tcp/ServerImpl.java]: create

File: modules/core/src/main/java/org/apache/ignite/spi/discovery/tcp/ServerImpl.java [hunk 5]
```diff
@@ -2070,6 +2073,27 @@ private void processMessageFailedNodes(TcpDiscoveryAbstractMessage msg) {
...
+        String warningMsg) {
```
Comment (NSAmelchev) [inline modules/core/src/main/java/org/apache/ignite/spi/discovery/tcp/ServerImpl.java]: warning

File: modules/core/src/main/java/org/apache/ignite/spi/discovery/tcp/ServerImpl.java [hunk 6]
```diff
@@ -2070,6 +2073,27 @@ private void processMessageFailedNodes(TcpDiscoveryAbstractMessage msg) {
...
+        TcpDiscoveryNodeFailedMessage nodeFailedMsg = new TcpDiscoveryNodeFailedMessage(
+            getLocalNodeId(), failedNode.id(), failedNode.internalOrder());
+
+        nodeFailedMsg.warning(warningMsg);
+
+        synchronized (mux) {
+            failedNodesWithMsgs.put(failedNode, warningMsg);
```
Comment (NSAmelchev) [inline modules/core/src/main/java/org/apache/ignite/spi/discovery/tcp/ServerImpl.java]: synchronize with failedNodes  map

File: modules/core/src/main/java/org/apache/ignite/spi/discovery/tcp/ServerImpl.java [hunk 7]
```diff
@@ -3217,6 +3245,9 @@ else if (e instanceof SocketTimeoutException ||
                     if (!failedNodes.contains(next)) {
                         failedNodes.add(next);
 
+                        failedNodesWithMsgs.put(next, "Node failed to send message to next node [msg=" + msg +
```
Comment (NSAmelchev) [inline modules/core/src/main/java/org/apache/ignite/spi/discovery/tcp/ServerImpl.java]: not synchronized

File: modules/core/src/main/java/org/apache/ignite/spi/discovery/tcp/ServerImpl.java [hunk 8]
```diff
@@ -3262,8 +3293,14 @@ else if (e instanceof SocketTimeoutException ||
                         failedNodesMsgSent.add(failedNode.id());
                 }
 
-                for (TcpDiscoveryNode n : failedNodes)
-                    msgWorker.addMessage(new TcpDiscoveryNodeFailedMessage(locNodeId, n.id(), n.internalOrder()));
+                for (TcpDiscoveryNode n : failedNodes) {
+                    TcpDiscoveryNodeFailedMessage failMsg = new TcpDiscoveryNodeFailedMessage(locNodeId, n.id(),
+                        n.internalOrder());
+
+                    failMsg.warning(failedNodesWithMsgs.get(n));
```
Comment (NSAmelchev) [inline modules/core/src/main/java/org/apache/ignite/spi/discovery/tcp/ServerImpl.java]: not synchronized

File: modules/core/src/main/java/org/apache/ignite/spi/discovery/tcp/ServerImpl.java [hunk 9]
```diff
@@ -5336,7 +5381,12 @@ private void checkFailedNodesList() {
                             if (msgs == null)
                                 msgs = new ArrayList<>();
 
-                            msgs.add(new TcpDiscoveryNodeFailedMessage(getLocalNodeId(), node.id(), node.internalOrder()));
+                            TcpDiscoveryNodeFailedMessage msg = new TcpDiscoveryNodeFailedMessage(getLocalNodeId(),
+                                node.id(), node.internalOrder());
+
+                            msg.warning(failedNodesWithMsgs.get(node));
```
Comment (NSAmelchev) [inline modules/core/src/main/java/org/apache/ignite/spi/discovery/tcp/ServerImpl.java]: maybe not synchronized

File: modules/core/src/main/java/org/apache/ignite/spi/discovery/tcp/ServerImpl.java [hunk 10]
```diff
@@ -207,6 +207,9 @@
     /** Failed nodes (but still in topology). */
     private final Map<TcpDiscoveryNode, UUID> failedNodes = new HashMap<>();
 
+    /** Failed nodes (but still in topology) associated with message.  */
+    private final Map<TcpDiscoveryNode, String> failedNodesWithMsgs = new HashMap<>();
```
Comment (NSAmelchev) [inline modules/core/src/main/java/org/apache/ignite/spi/discovery/tcp/ServerImpl.java]: maybe it better to create a single map? Where value as a tuple or class with info.

File: modules/core/src/main/java/org/apache/ignite/spi/discovery/tcp/ServerImpl.java (hunk 4 above)
Comment (SomeFire) [reply modules/core/src/main/java/org/apache/ignite/spi/discovery/tcp/ServerImpl.java]: Done.

File: modules/core/src/main/java/org/apache/ignite/spi/discovery/tcp/ServerImpl.java (hunk 5 above)
Comment (SomeFire) [reply modules/core/src/main/java/org/apache/ignite/spi/discovery/tcp/ServerImpl.java]: Done

File: modules/core/src/main/java/org/apache/ignite/spi/discovery/tcp/ServerImpl.java (hunk 10 above)
Comment (SomeFire) [reply modules/core/src/main/java/org/apache/ignite/spi/discovery/tcp/ServerImpl.java]: Impossible - in some cases they are used in different placeses.

File: modules/core/src/main/java/org/apache/ignite/spi/discovery/tcp/ServerImpl.java [hunk 11]
```diff
@@ -794,11 +798,7 @@ private void interruptPing(TcpDiscoveryNode node) {
         TcpDiscoveryNode node = ring.node(nodeId);
 
         if (node != null) {
-            TcpDiscoveryNodeFailedMessage msg = new TcpDiscoveryNodeFailedMessage(getLocalNodeId(),
-                node.id(),
-                node.internalOrder());
-
-            msg.warning(warning);
+            TcpDiscoveryNodeFailedMessage msg = newTcpDiscoveryNodeFailedMessage(node, warning);
```
Comment (NSAmelchev) [inline modules/core/src/main/java/org/apache/ignite/spi/discovery/tcp/ServerImpl.java]: create

File: modules/core/src/main/java/org/apache/ignite/spi/discovery/tcp/ServerImpl.java [hunk 12]
```diff
@@ -204,9 +204,12 @@
     @SuppressWarnings("FieldAccessedSynchronizedAndUnsynchronized")
     private StatisticsPrinter statsPrinter;
 
-    /** Failed nodes (but still in topology). */
+    /** Failed nodes (but still in topology) associated with detector(sender)'s id. */
     private final Map<TcpDiscoveryNode, UUID> failedNodes = new HashMap<>();
 
+    /** Failed nodes (but still in topology) associated with message.  */
+    private final Map<TcpDiscoveryNode, String> failedNodesMsgs = new HashMap<>();
```
Comment (NSAmelchev) [inline modules/core/src/main/java/org/apache/ignite/spi/discovery/tcp/ServerImpl.java]: maybe use a concurrent map instead mux?

File: modules/core/src/test/java/org/apache/ignite/spi/discovery/tcp/TcpFailMessagesTest.java [hunk 13]
```diff
@@ -0,0 +1,125 @@
...
+    /**
+     * i0 - another node, where additional info (warning message) should be logged.
+     * i1 - failure detector.
+     * i2 - failing node.
+     *
+     * @throws Exception If failed.
+     */
+    public void testFailMsgOnOtherNode() throws Exception{
+        final Ignite i0 = startGrid(0);
+        Ignite i1 = startGrid(1);
```
Comment (NSAmelchev) [inline modules/core/src/test/java/org/apache/ignite/spi/discovery/tcp/TcpFailMessagesTest.java]: why not final?

File: modules/core/src/test/java/org/apache/ignite/spi/discovery/tcp/TcpFailMessagesTest.java [hunk 14]
```diff
@@ -0,0 +1,125 @@
...
+
+        EventListener prd = new EventListener(latch);
+
+        i0.events().localListen(prd, EVT_NODE_FAILED);
+        i1.events().localListen(prd, EVT_NODE_FAILED);
+        i2.events().localListen(prd, EVT_NODE_SEGMENTED);
+
+        awaitPartitionMapExchange();
+
+        i1.configuration().getDiscoverySpi().failNode(i2.cluster().localNode().id(), "Forced fail.");
```
Comment (NSAmelchev) [inline modules/core/src/test/java/org/apache/ignite/spi/discovery/tcp/TcpFailMessagesTest.java]: I would make  "Forced fail." a final variable.

Comment (SomeFire) [reply modules/core/src/test/java/org/apache/ignite/spi/discovery/tcp/TcpFailMessagesTest.java]: Done.

File: modules/core/src/test/java/org/apache/ignite/spi/discovery/tcp/TcpFailMessagesTest.java (hunk 13 above)
Comment (SomeFire) [reply modules/core/src/test/java/org/apache/ignite/spi/discovery/tcp/TcpFailMessagesTest.java]: Final modifier removed because of no need.

**Pull request closed without merging**